controllers/spacebindingrequestmigration/spacebindingrequest_migration_controller.go

```diff
@@ -0,0 +1,182 @@
+	mur := &toolchainv1alpha1.MasterUserRecord{}
+	if err := r.Client.Get(ctx, murName, mur); err != nil {
+		if errors.IsNotFound(err) {
+			// mur was deleted
+			return reconcile.Result{}, nil
+		}
+		// error while reading MUR
+		return ctrl.Result{}, errs.Wrapf(err, "unable to get the bound MUR")
+	}
+
+	// skip workspace creator spacebinding
+	// the controller will convert only spacebindings created by system admins using the sandbox-cli.
+	// If the creator label on the space matches the usersignup reference on the MUR then this is the owner of the space
+	// and the spacebinding should not be migrated.
+	if mur.ObjectMeta.OwnerReferences == nil || len(mur.ObjectMeta.OwnerReferences) == 0 {
+		return ctrl.Result{}, errs.New("MasterUserRecord has no UserSignup owner reference")
+	}
+	// skip spaces with no creator label
+	if _, ok := space.Labels[toolchainv1alpha1.SpaceCreatorLabelKey]; !ok {
+		// let's log an error, since this should not happen in production
+		logger.Error(errs.New("space has no SpaceCreatorLabelKey set"), "the spacebindings for this space will not be migrated", "space name", space.Name)
+		return ctrl.Result{}, nil
+	}
+	usersignup := mur.ObjectMeta.OwnerReferences[0]
+	if space.Labels[toolchainv1alpha1.SpaceCreatorLabelKey] == usersignup.Name {
+		return reconcile.Result{}, nil
+	}
+
+	// get the spaceRole
+	spaceRole := spaceBinding.Spec.SpaceRole
+
+	// get member cluster name where the space was provisioned
+	targetCluster := space.Spec.TargetCluster
+	memberCluster, memberClusterFound := r.MemberClusters[targetCluster]
+	if !memberClusterFound {
+		return ctrl.Result{}, errs.New(fmt.Sprintf("unable to find member cluster: %s", targetCluster))
+	}
+
+	// get the home namespace from space
+	defaultNamespace := getDefaultNamespace(space.Status.ProvisionedNamespaces)
+
+	// construct a SpaceBindingRequest object
+	sbrName := mur.GetName() + "-" + spaceRole
+	sbr := &toolchainv1alpha1.SpaceBindingRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sbrName,
+			Namespace: defaultNamespace,
+		},
+		Spec: toolchainv1alpha1.SpaceBindingRequestSpec{
+			MasterUserRecord: mur.GetName(),
+			SpaceRole:        spaceRole,
+		},
+	}
+
+	// check if SpaceBindingRequest doesn't exist already
+	err = memberCluster.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: defaultNamespace,
+		Name:      sbrName,
+	}, &toolchainv1alpha1.SpaceBindingRequest{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// create the sbr resource
+			if err := memberCluster.Client.Create(ctx, sbr); err != nil {
+				return ctrl.Result{}, errs.Wrapf(err, "unable to create SpaceBindingRequest")
+			}
+
+			// let's requeue after we created the SBR, so that in next loop the migrated SpaceBinding object will be deleted
+			return ctrl.Result{Requeue: true}, nil
+		}
+		// Error reading the object
+		return ctrl.Result{}, err
+	}
```

Comment:
Suggestion: maybe use [the `CreateOrUpdate` function from `controllerutil`][1]?  Up to you if you want to leave it as-is, but it might make this part of reconciliation easier to follow.

[1]: https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.3/pkg/controller/controllerutil#CreateOrUpdate

Reply:
Nice suggestion. I haven't used this function before tbh, we have few use case where it could improve readability 🤔 
  
In this specific case, I don't actually need to update anything. 
In my case is important to know if the SBR already exists so I can delete the SpaceBinding (see line 163). If it doesn't exist I need to create it and force a reconcile of the same object ( see line 156 ) 
I guess I could still do this by checking the `op` returned if it's `created` (force reconcile) or one between `updated ` or `unchanged` (delete the SB). But still it wouldn't look obvious to me what is going on. WDYT?

Reply:
I refactored it to see how it'd turn out, and this is what I wound up with:

<details><summary>Reconcile method epilogue</summary>
<p>

```go

	// construct a SpaceBindingRequest object
	sbrName := mur.GetName() + "-" + spaceRole
	sbr := &toolchainv1alpha1.SpaceBindingRequest{
		ObjectMeta: metav1.ObjectMeta{
			Name:      sbrName,
			Namespace: defaultNamespace,
		},
	}

	result, err := controllerutil.CreateOrUpdate(ctx, memberCluster.Client, sbr, func() error {
		sbr.Spec = toolchainv1alpha1.SpaceBindingRequestSpec{
			MasterUserRecord: mur.GetName(),
			SpaceRole:        spaceRole,
		}
		return nil
	})

	if err != nil {
		// something happened when we tried to read or write the sbr
		return ctrl.Result{}, errs.Wrapf(err, "Failed to create or update space binding request %v", sbrName)
	}

	if result == controllerutil.OperationResultCreated {
		// let's requeue after we created the SBR, so that in next loop the migrated SpaceBinding object will be deleted
		return ctrl.Result{Requeue: true}, nil
	}

	// if the SBR was found (was created from the previous reconcile loop), we can now delete the SpaceBinding object
	if err := r.Client.Delete(ctx, spaceBinding); err != nil {
		return ctrl.Result{}, errs.Wrapf(err, "unable to delete the SpaceBinding")
	}

	return ctrl.Result{}, nil
}

```

</p>
</details> 

IMO it looks a bit nicer, but I'll leave it for you to decide.

If you're interested, here's the full diff:

<details><summary>The full diff</summary>
<p>

```diff
diff --git a/controllers/spacebindingrequestmigration/spacebindingrequest_migration_controller.go b/controllers/spacebindingrequestmigration/spacebindingrequest_migration_controller.go
index ae79c1d..c72df74 100644
--- a/controllers/spacebindingrequestmigration/spacebindingrequest_migration_controller.go
+++ b/controllers/spacebindingrequestmigration/spacebindingrequest_migration_controller.go
@@ -14,6 +14,7 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 	ctrl "sigs.k8s.io/controller-runtime"
 	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
@@ -134,32 +135,27 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
 			Name:      sbrName,
 			Namespace: defaultNamespace,
 		},
-		Spec: toolchainv1alpha1.SpaceBindingRequestSpec{
+	}
+
+	result, err := controllerutil.CreateOrUpdate(ctx, memberCluster.Client, sbr, func() error {
+		sbr.Spec = toolchainv1alpha1.SpaceBindingRequestSpec{
 			MasterUserRecord: mur.GetName(),
 			SpaceRole:        spaceRole,
-		},
-	}
-
-	// check if SpaceBindingRequest doesn't exist already
-	err = memberCluster.Client.Get(ctx, types.NamespacedName{
-		Namespace: defaultNamespace,
-		Name:      sbrName,
-	}, &toolchainv1alpha1.SpaceBindingRequest{})
-	if err != nil {
-		if errors.IsNotFound(err) {
-			// create the sbr resource
-			if err := memberCluster.Client.Create(ctx, sbr); err != nil {
-				return ctrl.Result{}, errs.Wrapf(err, "unable to create SpaceBindingRequest")
-			}
-
-			// let's requeue after we created the SBR, so that in next loop the migrated SpaceBinding object will be deleted
-			return ctrl.Result{Requeue: true}, nil
 		}
-		// Error reading the object
-		return ctrl.Result{}, err
+		return nil
+	})
+
+	if err != nil {
+		// something happened when we tried to read or write the sbr
+		return ctrl.Result{}, errs.Wrapf(err, "Failed to create or update space binding request %v", sbrName)
 	}
 
-	// if the SBR was found ( was created from the previous reconcile loop), we can now delete the SpaceBinding object
+	if result == controllerutil.OperationResultCreated {
+		// let's requeue after we created the SBR, so that in next loop the migrated SpaceBinding object will be deleted
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// if the SBR was found (was created from the previous reconcile loop), we can now delete the SpaceBinding object
 	if err := r.Client.Delete(ctx, spaceBinding); err != nil {
 		return ctrl.Result{}, errs.Wrapf(err, "unable to delete the SpaceBinding")
 	}

```

</p>
</details>

Reply:
Nice job @sadlerap . I agree with you, this indeed looks nicer and easier to read. It is indeed a pattern we should use more often in our code.

My only "concern" , so to speak, is about the mutateFn, which will trigger an update of the object, and an extra reconcile if I'm not wrong. Given that this controller will reconcile all the spacebindings CRs (800 objects for prod), and convert those to SBRs, there will be potentially 800 extra reconcile loops when we deploy it, it could be a performance concern and cause some slowness to the migration task. But I might be wrong.

Reply:
If it doesn't post anything to k8s in case of `unchanged` result, then that's good. 
But I wouldn't add the check for changed objects inside the function.

Let me have a closer look at what it does and update the code eventually. Thanks again for looking into it!

Reply:
@sadlerap I've implemented your first suggestion , please take a look at https://github.com/codeready-toolchain/host-operator/pull/922/commits/aaa038320bb7746cfb27eefa2ed5d9206eb90a9b 

thanks!